src/core/lease.ts

```diff
@@ -0,0 +1,235 @@
+    }
+
+    public async isExpired (): Promise<boolean> {
+        const lease = await this.retrieveLease()
+        return !!lease && Lease.expired(lease)
+    }
+
+    private async retrieveLease (): Promise<V1Lease> {
+        try {
+            return await this.client.readNamespacedLease(this.LeaseName, this.Namespace)
+        } catch (e: any) {
+            if (!(e instanceof SoloError)) {
+                throw new LeaseAcquisitionError(`failed to read the lease named '${this.LeaseName}' in the ` +
+                    `'${this.Namespace}' namespace, caused by: ${e.message}`, e)
+            }
+
+            if (e.meta.statusCode !== 404) {
+                throw new LeaseAcquisitionError('failed to read existing leases, unexpected server response of' +
+                    `'${e.meta.statusCode}' received`, e)
+            }
+        }
+
+        return null
+    }
+
+    private async createOrRenewLease (lease: V1Lease): Promise<void> {
+        try {
+            if (!lease) {
+                await this.client.createNamespacedLease(this.LeaseName, this.Namespace, this.LeaseHolder.toJson(), this.DurationSeconds)
+            } else {
+                await this.client.renewNamespaceLease(this.LeaseName, this.Namespace, lease)
+            }
+
+            if (!this.scheduleId) {
+                this.scheduleId = await this.renewalService.schedule(this)
+            }
+        } catch (e: any) {
+            throw new LeaseAcquisitionError(`failed to create or renew the lease named '${this.LeaseName}' in the ` +
+                `'${this.Namespace}' namespace`, e)
+        }
+    }
+
+    private async transferLease (lease: V1Lease): Promise<void> {
+        try {
+            await this.client.transferNamespaceLease(lease, this.LeaseHolder.toJson())
+
+            if (!this.scheduleId) {
+                this.scheduleId = await this.renewalService.schedule(this)
+            }
+        } catch (e: any) {
+            throw new LeaseAcquisitionError(`failed to transfer the lease named '${this.LeaseName}' in the ` +
+                `'${this.Namespace}' namespace`, e)
+        }
+    }
+
+    private async deleteLease (): Promise<void> {
+        try {
+            await this.client.deleteNamespacedLease(this.LeaseName, this.Namespace)
+        } catch (e: any) {
+            throw new LeaseRelinquishmentError(`failed to delete the lease named '${this.LeaseName}' in the ` +
+                `'${this.Namespace}' namespace`, e)
+        }
+    }
+
+    private static expired (lease: V1Lease): boolean {
+        const now = Date.now()
+        const durationSec = lease.spec.leaseDurationSeconds || Lease.DEFAULT_LEASE_DURATION
+        const lastRenewal = lease.spec?.renewTime || lease.spec?.acquireTime
+        const deltaSec = (now - new Date(lastRenewal).valueOf()) / SECONDS
+        return deltaSec > durationSec
+    }
+
+    private heldBySameProcess (lease: V1Lease): boolean {
```

Comment:
@leninmehedy Because PID's are only guaranteed to be unique for a given user and given machine, we have to check all three to be sure it is the same process. Two user or two different machines could have the same PID.